packages/contract/src/pay/DaimoPay.sol

```diff
@@ -0,0 +1,269 @@
+
+        // Finally, forward the balance to the current recipient
+        address recipient = intentToRecipient[address(intentContract)];
+        if (recipient == address(0)) {
+            // No LP showed up, so just complete the action.
+            recipient = intent.finalCall.to;
+
+            intentToRecipient[address(intentContract)] = recipient;
+            completeAction(address(intentContract), intent, calls);
+        } else {
+            // Otherwise, the LP fastFinished the action, give them the recieved
+            // amount.
+            intent.bridgeTokenOut.addr.safeTransfer(
+                recipient,
+                intent.bridgeTokenOut.amount
+            );
+        }
+
+        emit Claim({
+            intentAddr: address(intentContract),
+            finalRecipient: recipient,
+            intent: intent
+        });
+    }
+
+    // Swap bridgeTokenOut to finalCallToken
+    // Then, if the action has a finalCall, make the action calls.
+    // Otherwise, transfer the token to the action address.
+    function completeAction(
+        address intentAddr,
+        PayIntent calldata intent,
+        Call[] calldata calls
+    ) internal {
+        // Run arbitrary calls provided by the LP. These will generally approve
+        // the swap contract and swap if necessary
+        for (uint256 i = 0; i < calls.length; ++i) {
+            Call calldata call = calls[i];
+            (bool success, ) = call.to.call{value: call.value}(call.data);
+            require(success, "DP: swap call failed");
+        }
+
+        // Check that swap had a fair price
+        uint256 finalCallTokenBalance = TokenUtils.getBalanceOf(
+            intent.finalCallToken.addr,
+            address(this)
+        );
+        require(
+            finalCallTokenBalance >= intent.finalCallToken.amount,
+            "DP: insufficient final call token received"
+        );
+
+        if (intent.finalCall.data.length > 0) {
+            // If the intent is a call, approve the final token and make the call
+            TokenUtils.approve(
+                intent.finalCallToken.addr,
+                address(intent.finalCall.to),
+                intent.finalCallToken.amount
+            );
+            (bool success, ) = intent.finalCall.to.call{
+                value: intent.finalCall.value
+            }(intent.finalCall.data);
+
+            if (success) {
+                emit ActionCompleted({
+                    intentAddr: intentAddr,
+                    destinationAddress: intent.finalCall.to
+                });
+            } else {
+                TokenUtils.transfer(
+                    intent.finalCallToken.addr,
+                    payable(intent.refundAddress),
+                    intent.finalCallToken.amount
+                );
+
+                emit ActionBounced({
+                    intentAddr: intentAddr,
+                    refundAddress: intent.refundAddress
+                });
```

Comment:
in case where we combine `IntentFinished` / `FastFinish`, this function can return `boolean success`